`linux_zen` & `linux_lqx`: unify, add `updateScript` & `linux_lqx` -> 5.18.7

psydvl opened the pull request:
###### Motivation for this change
Unify linux_zen & linux_lqx into zenKernels
Add `updateScript`
[Linux 5.16.18](https://lwn.net/Articles/889437/)
`linux_lqx: 5.15.16-lqx2 -> 5.16.18-lqx1`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Atemu reviewed (changes requested):
Unify linux_zen and linux_lqx -> zen_kernels: unify drv format

Might also want to truly unify them in code as a function.

psydvl commented:
@Atemu I'm not sure, did I understand you correctly.

Atemu commented:
Please read Contributing.md w.r.t. commit messages.

psydvl commented:
@Atemu How about this?

Atemu commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -0,0 +1,46 @@
+{ lib, fetchFromGitHub, buildLinux, ... } @ args:
+
+let
+  version = "5.16.12";
```
The version won't always be equal.

Atemu commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -0,0 +1,46 @@
+    sha256 = "z/KCE+s2Wg9Wbo+QOYhbCQXyPlWpIyMORBmuZg3h50o=";
+    isLqx = true;
+  };
+in
+let
```
```suggestion
```

Atemu commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -0,0 +1,46 @@
+  };
+in
+let
+  fillVariant = variant@{version, suffix, ...}: variant // {
+    modDirVersion = "${version}-${suffix}";
+    branch = lib.versions.majorMinor version + "/master";
+  };
```
We don't need this abstraction, just inline these two where they're used.

Atemu commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -0,0 +1,46 @@
+    modDirVersion = "${version}-${suffix}";
+    branch = lib.versions.majorMinor version + "/master";
+  };
+  zenKernels = variant@{version, modDirVersion, sha256, branch, isLqx, ...}: buildLinux (args // {
```
```suggestion
  zenKernelFor = { version, sha256, isLqx }: buildLinux (args // {
```

Atemu reviewed (changes requested):
Needs some simplifications.

Atemu commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -0,0 +1,40 @@
+{ lib, fetchFromGitHub, buildLinux, ... } @ args:
```
I'm not sure this is right, args would apply to both kernels now? I think we should adopt a similar pattern to emacs versions.

psydvl replied:
It seems all works
```
nix-repl> linuxKernel.kernels.zenKernels.zen.kernelPatches
[ { ... } { ... } ]
nix-repl> linuxKernel.kernels.zenKernels.lqx.kernelPatches
[ { ... } { ... } ]
```

Atemu replied:
Problem is that the args might need to be different in the future. zen might be a version ahead and not need a patch anymore for example.

psydvl replied:
I don't understand you: how `args` need to be different in future?
With `buildLinux` derivation changed, all kernels derivation in `pkgs/os-specific/linux/kernel` will be changed too.
And what do you mean by patching?
`zen-kernels` not patching by us, we're already downloading patched version

Atemu replied:
I mean the kernelPatches that are applied to all of our kernels. There might be new ones for newer kernel versions or old ones get obviated. Since zen and lqx have different release cadences, the patches might need to be different aswell.

Atemu replied:
Ooh that's neat, didn't ever think of using partial function application with callPackage. That's definitely the way to go.

psydvl replied:
Should I make these changes or keep them for the future if we are faced with the need to make them?

Atemu replied:
I'd make them now. The kernels are somewhat separate from another, so it makes sense to declare them as such. Maybe add a comment to `zenKernels` to explain why it exists.

Atemu commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -0,0 +1,40 @@
+    sha256 = "YLc05sQ0LFoKDCzxMxZo7uD3H0uNIIo0cBNHT3CXVZo=";
+    isLqx = true;
+  };
+  zenKernelsFor = {version, suffix, sha256, isLqx}: buildLinux (args // {
```
```suggestion
  zenKernelsFor = { version, suffix, sha256, isLqx }: buildLinux (args // {
```

psydvl replied:
Done

Atemu reviewed (approved):
Small nit, otherwise LGTM.

Atemu commented:
Needs to be rebased onto https://github.com/NixOS/nixpkgs/pull/165524

psydvl committed:
linux_lqx: 5.15.16-lqx2 -> 5.18.7-lqx1

psydvl commented:
Rebased
Also reorder commits and add `updateScript`

andrevmatos commented on pkgs/os-specific/linux/kernel/zen-kernels.nix:
```diff
@@ -1,16 +1,19 @@
-    version = "5.17";
-    suffix = "zen1";
-    sha256 = "sha256-8DCcO2oPnwXjjm3PW7vNZKbB3S7XsEB/svAMhbPRJIo=";
+    version = "5.17"; #zen
```
This won't work in `modDirVersion` for `.0` patch versions. On these, zen tag/version is e.g. `5.17-zen1` (version will be detected as `5.17`), but [`actualModDirVersion` here](https://github.com/NixOS/nixpkgs/blob/80d8655c15055472f0118a877351085cc22c1e92/pkgs/os-specific/linux/kernel/manual-config.nix#L174-L177) will expect `5.17.0` and will error. That's why I have set version as the full `5.17.0-zen1` on the last update and special-cased `rev` to be the `v5.17-zen1` tag on this case.

psydvl replied:
Fixed

psydvl committed:
linux_lqx, linux_zen: refactor to unify
Unify linux_zen and linux_lqx -> zen-kernels

PedroHLC commented:
@psydvl any interest in rebasing, updating, and merging this? It still looks good to me.

psydvl committed:
linux_lqx, linux_zen: add update script

psydvl committed:
linux_zen: 5.18.5-zen1 -> 5.18.7-zen1

psydvl commented:
@PedroHLC 
Just rebased and updated to the latest versions
However, 
1) I think this was ready to merge long ago
2) I've moved from NixOS and can't fully check that kernel works

PedroHLC reviewed (approved):
About the `update-zen.py`

- Running it without arguments, results in `Exception: Update variant expected`;
- Running it with the wrong arguments, results in `Exception: Unexepected variant instead of 'zen' or 'lqx'`;
- Running it with `zen` takes some time but seems to work as expected.
